### ISSUE TITLE
a way to type libraries that may run in either nodejs or the browser

### ISSUE THREAD

**Issue by DetachHead** [microsoft/TypeScript#53308]
# Suggestion

## 🔍 Search Terms
node or browser types

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [X] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [X] This wouldn't change the runtime behavior of existing JavaScript code
* [X] This could be implemented without emitting different JS based on the types of the expressions
* [X] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [X] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion
i have a library that could either be run in nodejs or the browser. this means i don't want any global types to be available that are not present in both the nodejs and browser runtimes.

## 📃 Motivating Example
with only the dom types:
```ts
window // no error
process // error (only available in node)
setTimeout // no error (available in both node and browser)
```

with only node types:
```ts
window // error (only available in browser)
process // no error
setTimeout // no error (available in both node and browser)
```

basically i would like to define sort of a "union" of the browser and node types, so that i get an error when attempting to access something that's not available in both environments.

## 💻 Use Cases
- when developing a library that supports both browsers and nodejs

**Comment by Andarist** [comment]
It requires some setup but you can achieve this today with `package.json#imports`. You can isolate your env-specific things in per-environment files and import them using a single source, like:
```ts
import { foo } from '#isomorphic_foo'
```
and a `package.json` like:
```ts
{
  "isomorphic_foo": {
    "browser": "./browser_foo.js",
    "node": "./node_foo.js"
  }
}
```

Using this technique you could run `tsc` multiple times, with different `customConditions`, to validate correctness. The tricky part though is that you would probably have to rewrite this field before publishing. In your local setup, those should point to your `.ts` files while in the published package those should point to `.js` files. On top of that... you'd have to ensure that both "setups" (local and published ones) can load the types for those correctly. Alternatively, instead of rewriting `package.json` content before publishing you could locally use an additional `source` condition and "route" resolution to node/browser through it.

What I want to say is that it's possible today **but** quite involved and complicated 😬
